### ISSUE TITLE
Federated Sharing: second user can't edit the file a second time when encryption is enabled

### ISSUE THREAD

**Issue by jnfrmarks** [owncloud/core#16344]
### Steps to reproduce

1. Install 8.1 to two servers
2. on server 2, enable encryption
3. upload a file on server1 
4. Share the file with a user on sever2
5. the user on server 2 accepts the share and edits the file
6. the same user edits the file a second time.

### Expected behaviour

The user can edit the file a second time.

### Actual behaviour

The user is unable to access the file in any way - clicking on the file does not bring up the editor.  This error is thrown in the log file of server 2:

```

{"reqId":"Z6kFy1nldtI+UfEWeK+h","remoteAddr":"x.x.x.x,"app":"PHP","message":"fread(): Length parameter must be greater than 0 at \/var\/www\/html\/owncloud\/apps\/windows_network_drive\/3rdparty\/icewind\/streams\/src\/Wrapper.php#67","level":3,"time":"2015-05-13T19:17:05+00:00"}
{"reqId":"Z6kFy1nldtI+UfEWeK+h","remoteAddr":"x.x.x.x","app":"index","message":"Exception: {\"Exception\":\"OC\\\\Encryption\\\\Exceptions\\\\DecryptionFailedException\",\"Message\":\"Encryption library: Decryption (symmetric) of content failed: error:0607A082:digital envelope routines:EVP_CIPHER_CTX_set_key_length:invalid key length\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php(294): OCA\\\\Encryption\\\\Crypto\\\\Crypt->decrypt(false, false, 'BkLmBfYmtXY\\\/VBC...', 'AES-128-CFB')\\n#1 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php(275): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent('### Steps to re...', 'BkLmBfYmtXY\\\/VBC...', 'AES-128-CFB')\\n#2 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(432): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('### Steps to re...')\\n#3 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(278): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#5 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/encryption.php(173): stream_get_contents(Resource id #416)\\n#6 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(230): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Encryption->file_get_contents('')\\n#7 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(938): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->file_get_contents('')\\n#8 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(471): OC\\\\Files\\\\View->basicOperation('file_get_conten...', '\\\/\\\/pre_7.txt', Array)\\n#9 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(669): OC\\\\Files\\\\View->file_get_contents('\\\/\\\/pre_7.txt')\\n#10 \\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/files_texteditor\\\/ajax\\\/loadfile.php(39): OC\\\\Files\\\\Filesystem::file_get_contents('\\\/\\\/pre_7.txt')\\n#11 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(153) : runtime-created function(1): require_once('\\\/var\\\/www\\\/html\\\/o...')\\n#12 [internal function]: __lambda_func(Array)\\n#13 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(272): call_user_func('\\\\x00lambda_478', Array)\\n#14 \\\/var\\\/www\\\/html\\\/owncloud\\\/lib\\\/base.php(862): OC\\\\Route\\\\Router->match('\\\/apps\\\/files_tex...')\\n#15 \\\/var\\\/www\\\/html\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n#16 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php\",\"Line\":351}","level":4,"time":"2015-05-13T19:17:05+00:00"}

```

This message is in the server log file when the file is saved the first time:

```

{"reqId":"zTcFB9PjWgVbw7sGgpja","remoteAddr":"x.x.x.x","app":"webdav client","message":"curl GET http:\/\/172.18.5.74\/owncloud\/public.php\/webdav\/ returned status code 507","level":3,"time":"2015-05-13T19:16:22+00:00"}

```

### Server 1 Host info:

Ubuntu 14.04
PHP 5.5.9

### Server 2 Host info:

CentOS 6.6
PHP 5.4.36

### Server configuration:

ownCloud Enterprise Edition 8.1 beta 1 (daily) Build:2015-05-13T03:08:44+00:00 b626de0b9f135e2993496a8d18f3200dfab72337

**Comment by PVince81** [comment]
@schiesbn @th3fallen @DeepDiver1975

**Comment by DeepDiver1975** [comment]
@th3fallen please have a look - thx

**Comment by th3fallen** [comment]
was having some computer difficulties this morning, starting on it now.

**Comment by nickvergessen** [comment]
### Other steps to reproduce the issue:

1. Create two user
2. As user1 create file `test.txt`
3. As user1 share file `test.txt` with `user2@localhost`
4. As user2 accept the share
5. As user2 edit the file
6. As user1 open the file with the files_texteditor

### Expected
See the files content.

### Actual
The content is shown encrypted

### Additional fun

7. As user1 replace the content of the file with some text
8. As user2 try to open the file with the text_editor

### Expected
File is opened and the content is displayed

### Actual
![bildschirmfoto vom 2015-05-20 13 52 27](https://cloud.githubusercontent.com/assets/213943/7725074/98a44c4e-fef7-11e4-84e9-381ab7c2bfed.png)

Nothing in the log

**Comment by nickvergessen** [comment]
### Sync client problem

1. Create two user
2. As user1 create file `test.txt`
3. As user1 share file `test.txt` with `user2@localhost`
4. As user2 accept the share
5. Start a sync client for user2
6. Sync files for user2 with the client
7. As user2 edit the file
8. Sync files for user2 with the client

### Expected
The file is updated

### Actual
The file is kept as before and a conflict file is generated.

**Comment by schiesbn** [comment]
@th3fallen could you reproduce the issue? Do you already work on a solution?

**Comment by nickvergessen** [comment]
> Other steps to reproduce the issue

Works without encryption just fine

> Sync client problem

Indeed also happens without encryption, so I will create a new issue?

**Comment by th3fallen** [comment]
@schiesbn im able to duplicate it but haven't been able to solve it yet. I do know that user 1 would never be able to edit the file again since encryption is going to encrypt the file user2 edited with their keys and since user1 who is on the other server doesn't have encryption enabled he has no keys to encrypt it with.

So we need an exclusion in the wrapper is my guess there. As for the decryption issue im guessing it could be related to the file not being in that instances filesystem since its technically on the remote servers 


Thoughts on that @schiesbn

**Comment by nickvergessen** [comment]
@th3fallen just a side note: I used the same instance for both users on my test. (so both users having encryption enabled and having keys set up previous to the action)

**Comment by schiesbn** [comment]
@th3fallen yes, that's also what we did in ownCloud 8. We need to exclude all server-to-server mounts from encryption. The source server can encrypt the data but never the target server.

This is probably the right place to check if the storage is a https://github.com/owncloud/core/blob/master/lib/private/encryption/util.php#L357 "server-to-server share storage" and return true if this is the case: https://github.com/owncloud/core/blob/master/lib/private/encryption/util.php#L357

This is how the check looked in ownCloud <= 8.0. You can probably copy&paste it directly. https://github.com/owncloud/core/blob/stable8/apps/files_encryption/lib/proxy.php#L82

**Comment by nickvergessen** [comment]
since jennifer is on vacation atm, I will setup a second server, see whether the initial problems happens to me aswell and then try the patch and see if that fixes the stuff.

**Comment by nickvergessen** [comment]
Okay, adding ` && !($storage instanceof \OCA\Files_Sharing\External\Storage)` to https://github.com/owncloud/core/blob/master/lib/private/encryption/manager.php#L210 works for:

s1 | s2
--- | ---
enabled | enabled
disabled | enabled
enabled | disabled
disabled | disabled

Not sure if comparing to OCA is the best idea in terms of private/public api and depending on apps, but that's another topic.

I was also able to reproduce the problem from jennifer and it also get's resolved by the patch.

**Comment by schiesbn** [comment]
great! @th3fallen can you create a PR and also add a unit test for it? Thanks!

**Comment by PVince81** [comment]
@nickvergessen I think you need to use `$storage->instanceOfStorage()` to make it look through all the wrappers. Unless this was intended ?

**Comment by nickvergessen** [comment]
> @nickvergessen I think you need to use $storage->instanceOfStorage() to make it look through all the wrappers. Unless this was intended ?

Should happen on the other comparison aswell I guess, but yeah.

**Comment by schiesbn** [comment]
Not sure if it is needed, because at this place we apply a new wrapper to the storage so $storage should always be the base storage. But it also doesn't hurt to use the instanceOfStorage() method if it exists.

**Comment by th3fallen** [comment]
@schiesbn i don't have a patch yet still tracking down the actual culprit

**Comment by nickvergessen** [comment]
@th3fallen the patch is in his comment? Open the link and add the instanceof with the second class

**Comment by th3fallen** [comment]
Derp sorry missed that one @nickvergessen

**Comment by cmonteroluque** [comment]
@schiesbn @PVince81 @DeepDiver1975 any chance to review this PR?

**Comment by cmonteroluque** [comment]
@th3fallen can you merge the PR with the comments from @schiesbn today?

**Comment by davitol** [comment]
:+1: Fixed in version":"8.1.0.8","versionstring":"8.1 RC2"
